lte/gateway/c/session_manager/StatsPoller.cpp

```diff
@@ -13,18 +13,42 @@
 #include <stdint.h>
 #include <atomic>
 #include <memory>
+#include <vector>
 #include "StatsPoller.h"
-#define COOKIE 0
-#define COOKIE_MASK 0
+#define OVS_COOKIE_MATCH_ALL 0xffffffff
 
 namespace magma {
 
 void StatsPoller::start_loop(
     std::shared_ptr<magma::LocalEnforcer> local_enforcer,
-    uint32_t loop_interval_seconds) {
+    std::shared_ptr<magma::ShardTracker> shard_tracker) {
+  // when no shard ids are available poll all UEs
+  // to poll by cookie pass in a vector of shard ids, where
+  // each element is a shard id that maps directly to a cookie
   while (true) {
-    local_enforcer->poll_stats_enforcer(COOKIE, COOKIE_MASK);
-    std::this_thread::sleep_for(std::chrono::seconds(loop_interval_seconds));
+    std::vector<unsigned int> active_shard_ids =
+        shard_tracker->get_active_shards();
+    uint32_t interval;
+    // Set interval to be per shard. Here the maximum amount of shards is
+    // ten, but this can be flexible
+    if (active_shard_ids.size() == 0) {
+      interval = 10;
+    } else {
+      interval = 10 / active_shard_ids.size();
+    }
+    // Poll all stats when there are no active shards
+    // otherwise poll each shard id between intervals
+    if (active_shard_ids.size() == 0) {
+      local_enforcer->poll_stats_enforcer(0, 0);
+      std::this_thread::sleep_for(std::chrono::seconds(interval));
+    } else {
+      for (size_t shard_index = 0; shard_index < active_shard_ids.size();
+           shard_index++) {
+        local_enforcer->poll_stats_enforcer(
+            active_shard_ids[shard_index], OVS_COOKIE_MATCH_ALL);
+        std::this_thread::sleep_for(std::chrono::seconds(interval));
+      }
+    }
```

Comment:
Could we clean up this logic to look like 
```
if (active_shard_ids.size() == 0) {
// poll everything
// sleep for 10 seconds
// continue
}

// Do the else clause
```
?

Reply:
Also, if there are no active shards `active_shard_ids.size() == 0`, why do we need to poll at all?

Reply:
Also, let's use `active_shard_ids.empty()`

Reply:
So this is a change to see if it optimizes polling. Polling everything and sleep for 10 seconds was how I did polling before. The new idea was to reduce load by polling a shard every few seconds rather than polling at once. I guess technically we don't need to poll when there are no active shard ids.